### DISCUSSION TITLE
Custom Switch Activity

### DISCUSSION THREAD

**Discussion by Juandavi1** [elsa-workflows/elsa-core#1299] [Q&A] [answered]
Hi, 

we want to create a custom switch activity with a definition more friendly like a fork activity with the tags but the designer doesnt show the cases.

try defining the attribute as follows:

<img width="1113" alt="Captura de Pantalla 2021-07-22 a la(s) 12 08 39 a  m" src="https://user-images.githubusercontent.com/18731612/126592432-f38b16aa-39da-43d0-81fa-f4bc63a9503a.png">

result: 
<img width="935" alt="Captura de Pantalla 2021-07-22 a la(s) 12 13 20 a  m" src="https://user-images.githubusercontent.com/18731612/126592654-b7ba3d48-a170-4bc6-9fb2-8908655050db.png">

<img width="493" alt="Captura de Pantalla 2021-07-22 a la(s) 12 22 24 a  m" src="https://user-images.githubusercontent.com/18731612/126593262-3eafe833-8d90-4c8a-b8cb-3eed8a114c4e.png">

do you know how we can do it ? 

Thank you so much !!

**Comment by sfmskywalker** [comment, marked as answer]
Hi @Juandavi1,

You can do this by implementing a custom designer plugin that takes the tags the user enters and use them as outcomes for the activity.
Here's an example:

```html
<script>
    // Define a plugin.
    function AnswerTreePlugin(elsaStudio){
        const eventBus = elsaStudio.eventBus;

        // Register a handler for the 'activity-design-displaying' event.
        eventBus.on('activity-design-displaying', this.onActivityDesignDisplaying);

        this.onActivityDesignDisplaying = function(context){
            const activityModel = context.activityModel;

            // Only handle activities of type "AnswerTree".
            if (activityModel.type !== 'AnswerTree')
                return;

            const props = activityModel.properties || [];
            const syntax = 'Json';
            
            // Find the "Cases" property. If not found, initialize a new object with an empty array. 
            const casesProp = props.find(x => x.name === 'Cases') || {expressions: {'Json': '[]'}, syntax: syntax};
            const expression = casesProp.expressions[syntax] || [];
            
            // This is a bit rough on the edges still, but expressions can be serialized differently depending on certain scenarios.
            const cases = !!expression['$values'] ? expression['$values'] : Array.isArray(expression) ? expression : JSON.parse(expression) || [];
            
            // Set the outcomes of the activity context to the cases array.
            context.outcomes = cases; 
        }
    }

    // Get a handle to the elsa-studio-root element.
    const elsaStudioRoot = document.querySelector('elsa-studio-root');

    // Configure Elsa.
    (async () => {
        // Wait until the component is available.
        await customElements.whenDefined('elsa-studio-root');

        // Install plugins.
        await elsaStudioRoot.addPlugin(AnswerTreePlugin);
    })();
</script>
```

**Comment by Juandavi1** [comment]
Hi @sfmskywalker, 

Knowing this level of extensibility makes me appreciate this project even more. Thank you.

I'm trying to do what you told me to do but it seems that there is no addPlugin function inside elsaStudioRoot

<img width="539" alt="Captura de Pantalla 2021-07-22 a la(s) 11 44 15 p  m" src="https://user-images.githubusercontent.com/18731612/126738930-d7f00888-3eb4-4b21-aa6d-246fac8e5488.png">

<img width="697" alt="Captura de Pantalla 2021-07-22 a la(s) 11 44 25 p  m" src="https://user-images.githubusercontent.com/18731612/126738934-3b99cc0f-c611-44a8-939a-7ab76ef83525.png">
<img width="836" alt="Captura de Pantalla 2021-07-22 a la(s) 11 41 12 p  m" src="https://user-images.githubusercontent.com/18731612/126738936-6d13b4c0-fbf2-4470-9e05-7994dcb95504.png">

**Reply by sfmskywalker** [reply]
It’s a fairly new addition, so perhaps it’s not part of your version yet. Are you using prerelease packages or packages from NuGet?

**Reply by Juandavi1** [reply]
we are working with version 2.1.0.443 from mynuget

**Reply by sfmskywalker** [reply]
That would explain it. If you upgrade to the latest preview packages then you will be able to use the new functions.

If you prefer to stay on the stable releases then it will be available when 2.2 is released next month.

**Reply by Juandavi1** [reply]
thanks, i will try it

**Reply by Juandavi1** [reply]
Hi @sfmskywalker, 

the error disappeared but now the dashboard does not load :(

<img width="2251" alt="Captura de Pantalla 2021-07-23 a la(s) 1 24 28 p  m" src="https://user-images.githubusercontent.com/18731612/126825469-35790609-6544-4008-bac3-fdacf8c8303b.png">

<img width="902" alt="Captura de Pantalla 2021-07-23 a la(s) 1 32 40 p  m" src="https://user-images.githubusercontent.com/18731612/126826190-5f54f68a-d70b-4f5c-aa45-00ac3c7198b0.png">


I tried importing tailwind direct from cdn but it doesn't work for me either.

Thanks for the help you can give me.

**Reply by sfmskywalker** [reply]
Take a look at your _host.cshtml file and compare its contents with that of the Dashboard project in the master branch. It’s likely using incorrect script and style urls.

**Reply by Juandavi1** [reply]
Awesome !! I can see! 

Unfortunately the function registered through addPlugin is never called.

Do you know what it could be?

<img width="1406" alt="Captura de Pantalla 2021-07-23 a la(s) 2 05 20 p  m" src="https://user-images.githubusercontent.com/18731612/126829591-3ab1b0a0-1b13-47e3-8840-92f1ad587318.png">

**Reply by sfmskywalker** [reply]
Can you show me the full context of that file? Might be something small.

**Reply by Juandavi1** [reply]
<img width="1484" alt="Captura de Pantalla 2021-07-23 a la(s) 9 59 56 p  m" src="https://user-images.githubusercontent.com/18731612/126855614-1e7f946e-303a-43cd-b102-45783c814d21.png">

**Reply by Juandavi1** [reply]
this is the only configuration we have made

**Reply by sfmskywalker** [reply]
That all looks good to me. I don’t know why then it doesn’t invoke your plugin, since it works fine when I try.

The only thing I can think of next is that you send me a trimmed down version of a project that I can try for myself. If I can repro I can fix.

**Reply by Juandavi1** [reply]
perfect, I'm going to create a new project with just the plugin and I'll share it with you. Thanks a lot

**Reply by Juandavi1** [reply]
Hi @sfmskywalker, 

I added the plugin but now it won't let me add activities to the flow. I commented on this in the following issue https://github.com/elsa-workflows/elsa-core/issues/1313.
The function registered with AddPlugin is not invoked. I get the following error:

![Captura de pantalla de 2021-07-27 19-42-28](https://user-images.githubusercontent.com/18731612/127245635-771eb836-87cd-4844-9360-ff7a82956f7e.png)

**Reply by Juandavi1** [reply]
Hi @sfmskywalker 

I changed the definition of the function and now the function plugin is called and the cases are loaded.

before: 

![Captura de pantalla de 2021-07-27 20-23-35](https://user-images.githubusercontent.com/18731612/127248903-b8eeb12b-d204-4694-be12-6539fbb09e46.png)

now: 

![Captura de pantalla de 2021-07-27 20-25-09](https://user-images.githubusercontent.com/18731612/127248912-8e60be0a-5125-45e8-aa80-3a4e3c4f77ab.png)

Thanks

**Reply by martinjakubeci** [reply]
I am trying to add the plugin as described here, my _Host.cshtml is the same and version of packages is 2.2.0-preview.529.
I see this error when opening the designer: Uncaught (in promise) TypeError: Cannot read property 'addPlugin' of null
Any idea? @Juandavi1 how did you make it work please?

**Reply by martinjakubeci** [reply]
Finally made it work. However, the plugin seems to be working only when adding a new activity. After showing the workflow designer with an existing activity it shows the default "Done" outcome. No errors.

**Reply by sfmskywalker** [reply]
Is that still the case with the latest version of the code or package? There was a bug where event handlers would trigger only once, but should be good now. If not, let me know.

**Reply by Juandavi1** [reply]
It works ! Thanks a lot !
